rero_ils/modules/stats/api/report.py

```diff
@@ -0,0 +1,445 @@
+                    calendar_interval='year', format='yyyy'
+                ),
+                'birth_year': lambda: A(
+                    'date_histogram', field='birth_date',
+                    calendar_interval='year', format='yyyy'
+                ),
+                'postal_code': lambda: A(
+                    'terms', field='postal_code', size=self.aggs_size),
+                'gender': lambda: A(
+                    'terms', field='gender', size=self.aggs_size),
+                'role': lambda: A(
+                    'terms', field='roles', size=self.aggs_size),
+                'type': lambda: A(
+                    'terms', field='patron.type.pid', size=self.aggs_size),
+            },
+            'get_key': {
+                'created_month': lambda bucket: bucket.key_as_string,
+                'created_year': lambda bucket: bucket.key_as_string,
+                'type': lambda bucket:
+                f'{self.patron_types[bucket.key]} ({bucket.key})',
+                'birth_year': lambda bucket: bucket.key_as_string,
+                'gender': lambda bucket: bucket.key,
+                'postal_code': lambda bucket: bucket.key,
+                'role': lambda bucket: bucket.key,
+            }
+        }
+        base_query = PatronsSearch()[:0].filter(
+            'term', organisation__pid=self.org_pid)
+
+        if active:
+            def active_patron_query():
+                """The based patron query and filter by active pid."""
+                range_period = self._get_range_period(self.period)
+                active_patron_pids = [
+                    hit.loan.patron.pid for hit in LoanOperationLogsSearch()
+                    .source('loan.patron.pid')
+                    .get_logs_by_trigger(
+                        triggers=[
+                            ItemCirculationAction.EXTEND,
+                            ItemCirculationAction.REQUEST,
+                            ItemCirculationAction.CHECKIN,
+                            ItemCirculationAction.CHECKOUT
+                        ],
+                        date_range=range_period)
+                    .filter('term', organisation__value=self.org_pid).scan()
+                ]
+                return base_query.filter('terms', pid=active_patron_pids)
+
+            patron_cfg['query'] = active_patron_query
+        else:
+            patron_cfg['query'] = lambda: base_query
+        return patron_cfg
+
+    def _circulation_config(self, trigger):
+        """The es configuration for circulation indicators."""
+        return {
+            'query': lambda: (
+                s := RecordsSearch(index=LoanOperationLog.index_name)[:0]
+                .filter('term', organisation__value=self.org_pid)
+                .filter('term', record__type='loan')
+                .filter('term', loan__trigger=trigger),
+                s.filter('range', date=self._get_range_period(self.period))
+            )[self.period is not None],
```

Comment:
I don't understand all the logic use here.... Maybe a comment could be useful ?

Reply:
Done